Client-side specimen ID matching

fedarko opened the pull request:
Progress on #520 / #553. (Both of those issues have a couple moving parts, but this should address what I think is the "main" thing with both.)

## Summary of changes
- Added `getSubstringMatches()`, a simple function that (given a query string and an array of strings) returns all strings in the array that include the query string
  - Added corresponding unit tests
- Added `get_active_samples()` (named analogously to the existing `get_active_studies()`). This function just gets all sample IDs associated with all active studies. (It technically returns a `Promise` that resolves to an array of sample IDs, since the requests are done asynchronously.)
  - Made some modifications in the back-end code to enable this -- in particular, exposing a `limit` parameter that allows the front-end to issue a request for *all* samples associated with a study instead of just 20.
  - Added various tests to verify that `limit`s are being handled properly. 
  - Thanks to @charles-cowart for help with this!
  - Also created `merge_sample_responses()`, which is just a few lines of code I abstracted out from `autocomplete_search_samples()` so that I could use it in both that function and in `get_active_samples()`.
- Added an "error" function in `autocomplete_search_samples()` (which didn't previously have one). This function just creates a `bootstrapAlert()` explaining the situation.
  - Also added two other error functions to the two other instances in `plateViewer.js` that use `.then()` (i.e. Promises).
  - See the commit message of https://github.com/biocore/LabControl/pull/585/commits/ed55343e6c9855d92db0c9f0383fbefdbbcdfd86 for more details about this than are probably necessary.

- `PlateViewer.modifyWell()` now uses `get_active_samples()` and `getSubstringMatches()` to perform client-side matching: given an input specimen ID, this tries to match it up with an active sample ID.
  - There are three possible results of this:
    - exactly 1 sample matches --> great, change the well contents accordingly
    - 0 samples match --> ok, this well will just be labelled "unknown"
    - 2 or more samples match --> assign the new class "indeterminate" (it's the yellow one) to this well, indicating that manual resolution is needed. **On reloading a plate, these wells will show up as red (i.e. "unknown")** -- see caveats below. 
  - The main AJAX call that `PlateViewer.modifyWell()` used to do is now done in `PlateViewer.patchWell()`. I initially did this so I could call `patchWell()` from multiple points, but this ended up not being needed. (I still like having `modifyWell()` and `patchWell()` be separate functions, though. Feel free to ask for them to be renamed :)
  - Changed the "intermediate" well legend to reflect the behavior on reloading.
- Various small fixes around the codebase (adding missing docs in process.py, removed an unused variable declaration in plate.js, added Travis/Coveralls badges to the README, etc.)

## Caveats
There are some things I still think could be improved in the proposed changes -- noting them here because most of these are sorta subjective, so if you have any feedback I'd appreciate it.
- Client-side matching is fairly quick (at least on my computer), but it could be faster if we did stuff like caching the list of active sample IDs. The golden standard here would be applying all of these changes in bulk (instead of doing everything through `modifyWell()`, which is why you see samples' statuses change in individual "blips")... but this would probably necessitate some significant restructuring of how the plating interface works.

- ~~Some of the uses of promises in the code could be handled better (e.g. throwing a bootstrapAlert if a promise fails to resolve).~~ **Update: resolved as of https://github.com/biocore/LabControl/pull/585/commits/ed55343e6c9855d92db0c9f0383fbefdbbcdfd86.**

- This is the main caveat: after spending some time going through the code that handles well classes, it seems like adding an indeterminate class to wells on the front-end is pretty simple (as is done here). However, making that class persist on the back-end (so that reloading a plate reloads the class for each well) is going to be a pretty heavy amount of work on top of what's been done here: it seems like we'd need to modify the database schema to store this information, or at least figure out a way to use a SQL query to detect *in python* which wells are indeterminate. Took a few stabs at this last night; didn't get very far.
  - To complicate things, indeterminate wells are technically "unknown", since they don't have a valid value. From what I can tell, making these wells not "unknown" seems like it'd involve a lot of going through the code to modify things.
  - For now, the behavior is that (as mentioned above) the indeterminate class isn't preserved when reloading a plate—indeterminate cells will just show up on a reload as "unknown", which is *technically* kinda accurate since they don't contain exact sample IDs. Although technically accurate, this is fairly undesirable. (As mentioned, I updated the indeterminate legend to match, so that users understand why these cells change on reloading.)
  - I'm sure making the indeterminate class persistent is doable—and I'm happy to work when I get back next week on this—but I'm not sure that the effort is worth it, especially since the yellow cells are an indicator of "hey you need to resolve this manually". Of course if you disagree that is fine, I'm not the expert here :)

Anyway—sorry if this PR isn't comprehensive enough; I mainly wanted to make sure I had this ready to show you all where this is at today. Let me know what you think!

<img width="1552" alt="Screen Shot 2019-08-30 at 8 34 02 AM" src="https://user-images.githubusercontent.com/4177727/64020999-fc9b8580-cb00-11e9-9e52-da2ae8771784.png">

fedarko committed:
ENH: Expose limit in StudySamplesHandler #520 #553

This is an important part of implementing client-side multi-matching
(i.e. getting all the IDs needed to perform matching).

Pair-programmed with @charles-cowart :)

fedarko committed:
MAINT: add get_active_samples(); abstract code

get_active_samples() is based on autocomplete_search_samples(), so
I abstracted some of the code for processing responses into its own
function that can be shared.

This seems to work alright! Would really like to test this with
multiple studies, though (#437).

For future reference: you can get the value of a promise returned by
get_active_samples() as follows:

get_active_samples().then(
    function(v) {
        // Now, v is the array of active sample IDs
        console.log(v);
    }
);

This is 1/3 of the work towards multi-select (just gonna tag #520).

What remains are the following:
2. On a paste operation when the multi-select checkbox is checked,
   identify all "new" specimen IDs. Get an array or something of
   these, preferably. (Or we can do this on a
   specimen-ID-by-specimen-ID basis, doesn't really matter.)
3. For each new specimen ID, attempt to match it with the active
   samples. Color cells / update values accordingly based on the
   following cases:
   *   0 matches (color cell red)
   *   1 match (leave cell uncolored, and change text to matching ID)
   * >=2 matches (color cell yellow)

And ...I think that should be it, knock on wood.

fedarko committed:
Merge branch 'master' of https://github.com/jdereus/LabControl

just syncing with upstream

fedarko committed:
Merge branch 'master' of https://github.com/jdereus/LabControl

fedarko committed:
Merge branch 'master' of https://github.com/jdereus/LabControl

fedarko committed:
ENH: Prototype impl. of multi-select #520 #553

TODOs:

-Better then() failure handling on get_active_samples()
-Apply yellow well color (well-indeterminate CSS class IIRC) if multiple
 matches are possible.
-Remove debug console.log() stmts

fedarko committed:
MAINT: remove now-unneeded variable declaration

I think this was from back before I fixed #567

fedarko committed:
MAINT: Apply auto-matching regardless of ms c.box

Also prettified the code, since I forgot to do that previously.

fedarko committed:
Merge branch 'master' of https://github.com/jdereus/LabControl

fedarko committed:
MAINT: pass active study ID to patchWell

this avoids having to call getActiveStudy() twice, + removes
avenue for potential funky error where multiple calls would trigger
different results

fedarko committed:
Merge branch 'master' of https://github.com/jdereus/LabControl

fedarko committed:
DOC: Add comments re: multi-matching details

this includes some discussion of todos, etc

fedarko committed:
ENH: show yellow box on indet. wells *initially*

Right now it works ok if you don't reload the well. but ideally
this should save the indeterminate state on the backend? Which means
that I think I'll have to pretty much redo this js logic in python..?

fedarko committed:
DOC: Add TODO comment re: patchWell efficiency

tldr: it's slow, and could be faster

fedarko committed:
DOC: add comment re: a formerly-missing return val

fedarko committed:
HACK: Document how the indet. color is temporary

This change can definitely be ignored/reverted if desired, but this
makes it clear that if the user doesn't resolve this then these cells
will turn red when you reload the plate.

fedarko committed:
MAINT: Split getSubstringMatches into lc.js & test

fedarko committed:
MAINT: remove old console.log statements

fedarko committed:
MAINT: Add onRejected funcs throughout plateViewer

These will alert the user if something goes wrong. This sort of
paradigm is already in use in this codebase, so adding these is pretty
simple -- essentially just giving a short summary of the issue,
then dumping the error message.

fedarko committed:
MAINT: show responseText when applicable in alerts

This necessitated some messing around with each of the three
onRejected functions I added in 77cb5a6 in order to figure out which
arguments were passed where.

ERROR FUNCTIONS 1 AND 2

For the most part, any error functions that fail due to requests (i.e.
the error functions I just created in autocomplete_search_samples() and
in get_active_samples()) are passed a jqXHR, etc. as arguments -- so we
can just show the responseText.

(These error functions, if ever called,
should be called with a jqXHR since all they're really doing is waiting
on a series of requests to finish. I *guess* it's conceivably possible
they could fail not due to a request if jQuery's code somehow explodes,
but I doubt it.)

ERROR FUNCTION 3

The callback on the promise returned by get_active_samples() -- that
is, the error function declared in modifyWell() -- just receives a
normal rejectionReason as a string if there's an error in the success
function within get_active_samples(), which makes sense since if any
of the requests would fail in get_active_samples() then that would
trigger get_active_samples()' declared error function (with a jqXHR,
etc.  as arguments since a request caused the failure) instead of the
error function on the promise returned by get_active_samples().

TLDR

Phew! That was a lot of long paragraphs. The gist of things here is:
I tested each of the 3 new error callbacks locally, saw what the
responses I was getting were, and adjusted the bootstrapAlert() calls
accordingly to be more informative.
From manual testing and thinking things through these adjustments
should be correct, but even if they aren't (and, say,
error function 1 gets passed a string instead of a jqXHR and things)
then the outcome is the same: an error pops up with some sort of
message. (The difference in that case will be we'll see something like
"[Nice error message:] undefined" instead of
"[Nice error message:] [further information]", which isn't too
horrifying.)

fedarko commented:
As of now, I think this is ok to be reviewed—now that error handling is added in, the two remaining "caveats" are general, subjective issues that I'd like feedback on.

AmandaBirmingham commented:
@fedarko Hiya--I'm stepping into Charlie's shoes for the moment :)  I see that the Travis build for this PR did not succeed due to three failing unit tests.  Please modify the PR to address these failures; I can do a review on it once the build passes.

Thanks for all the effort you are putting in on LabControl!

fedarko committed:
TST: Fix test_get_study_samples_handler re changes

Adjusting tests re: the default limit no longer being 20.
Also adding some more tests now that "limit" is an exposed parameter.

A TODO is adding tests to labcontrol/db/tests/test_study.py that
check this functionality more closely (e.g. trying negative limits).

fedarko committed:
TST: Add missing response code check in a test

fedarko committed:
TST: Explicitly test & improve study samples limit

Previously in this PR, this wasn't directly tested, and things were
kind of funky -- Study.samples() said that it expected limit to be an
int, but the limit obtained from get_argument() in
StudySamplesHandler.get() was actually a str.

Major changes:

- Study.samples() now explicitly requires that limit is already an
  int.

- Added test_samples_with_limit() to TestStudy accordingly.

- StudySamplesHandler.get() now converts limit to an int (if it's been
  specified), before calling Study.samples(). If int() fails, then
  it'll raise a ValueError. (Similarly, Study.samples() raises a
  ValueError for invalid limits.) If StudySamplesHandler.get() runs
  into a ValueError, it sets a status code of 400 ("Bad Request").

- Added various invalid limit tests to
  test_get_study_samples_handler() accordingly.

Running tests on my system seems to explode the database on my system
(I'll look into getting this fixed soon -- since I've been working on
the front-end mostly this hasn't been a problem), so I'm going to just
kick this up to Travis for the time being.

fedarko committed:
TST: Fix error in exp output + beef up test cases

For some reason, I forgot that there were 9 "SKM" samples in the
test dataset. The thing I was trying to test with the

response = self.get('/study/1/samples?term=SKM&limit=30')

test was that the property of limits not doing anything if greater
than the normal amount of output still worked, even if a term was
specified. This test should be fixed now.

fedarko commented:
Thanks so much @AmandaBirmingham! This was due to me not updating tests re: the backend changes made in this PR. I updated the tests (and added some tests that the `limit` parameter sent to `Study.samples()` is being handled/validated properly), and things seem to be working on Travis now.

fedarko committed:
DOC: Add Travis badge to README [ci skip]

It might be worth setting this up with Codecov soon -- I'm curious as
to what particular spots of the code aren't covered.

fedarko committed:
DOC: whoops, add a linebreak after Travis badge

[ci skip]

fedarko committed:
DOC: Add Coveralls badge to README

fedarko commented:
Something potentially gnarly, but likely out of the scope of this PR... I ran ea536f3's build twice ([this one](https://travis-ci.org/biocore/LabControl/builds/581932706)), and it passed the first time but failed on the two "download" tests (`test_download` and `test_get_download_normalization_process_handler`) again.

Since these failures were both caused by `AssertionError: Async operation timed out after 10 seconds`, it seems like this is an intermittent error due to network problems (???). Not sure how best to handle this, but as a brute-force solution using `travis_retry` with the python tests might work.

fedarko committed:
BUG: Remove labcontrol/gui/test* from coverage

So it was omitting */tests*, but not */test* ...

Anyway, this will likely lower coverage by a fair amount, but it
should be a more accurate representation of codebase coverage.

gwarmstrong commented:
> Something potentially gnarly, but likely out of the scope of this PR... I ran [ea536f3](https://github.com/biocore/LabControl/commit/ea536f3faf559bce9b7c440b5ac6b493fce40de6)'s build twice ([this one](https://travis-ci.org/biocore/LabControl/builds/581932706)), and it passed the first time but failed on the two "download" tests (`test_download` and `test_get_download_normalization_process_handler`) when run again.
> 
> Since these failures were both caused by `AssertionError: Async operation timed out after 10 seconds`, it seems like this is an intermittent error due to network problems (???). Not sure how best to handle this, but as a brute-force solution using `travis_retry` with the python tests might work.

You're right, this has been a recurrent problem I've noticed in the past couple weeks. I think there are a couple things we can do to try to reduce this.
* Increase the `timeout` limit in the line below:
https://github.com/biocore/LabControl/blob/1427ad163682ff11b1086e0e3b1010ba1a2166ee/labcontrol/gui/testing.py#L71
* add retry functionality _within_ the test, e.g. something like [this](https://stackoverflow.com/questions/31529151/python-3-urllib-urlretrieve-retry-on-download-fail):
```
class TestDownloadLibraryPrepShotgunProcessHandler(TestHandlerBase):
    def test_download(self):
        retries = 0
        response = None
        while retries < 5:
            try:
                response = self.get(
                    '/process/library_prep_shotgun/%d/echo_pick_list' % 1)
            except AssertionError:
                print("Error downloading on try {}.".format(retries))
            else:
                break
        if response is None:
           raise AssertionError("Async operation timed out after maximum number of retries.")

        self.assertNotEqual(response.body, '')
        self.assertTrue(response.body.startswith(
            b'Sample ID\tSource Plate Name\t'))
        self.assertEqual(response.headers['Content-Disposition'],
                         "attachment; filename=2017-10-25_"
                         "Test_compressed_gDNA_plates_1-4_indices.txt")

```

AmandaBirmingham reviewed (changes requested):
Lots of good stuff here--thank you!  A few change requests, all pretty minor, I think.

fedarko commented on .coveragerc:
```diff
@@ -2,13 +2,15 @@
 branch = True
 omit =
     */tests*
+    */test*
```
Yeah, we need both. Essentially this file is saying "don't include coverage information for these directories in the total summary"—it was already excluding the `tests/` directory (i.e. `labcontrol/db/tests/`, but it wasn't excluding the `test/` directory for the front-end (`labcontrol/gui/test/`). This change should fix that problem, and the result is that coveralls should now provide a more realistic representation of the code coverage.

...It might also be worth renaming either of the test directories to make them consistent, since having the test directories be named slightly differently is a little bit confusing :)

gwarmstrong replied:
> ...It might also be worth renaming either of the test directories to make them consistent, since having the test directories be named slightly differently is a little bit confusing :)

I agree, opening an issue.

fedarko replied:
Marking this as resolved—we can follow up with this in #589.

fedarko commented on labcontrol/gui/js_tests/labcontrol_tests.js:
```diff
@@ -79,3 +79,32 @@ QUnit.module("clippingForPlateType", function(hooks) {
+
+QUnit.module("getSubstringMatches", function(hooks) {
+  QUnit.test("common usage", function(assert) {
+    assert.propEqual(getSubstringMatches("w", ["wahoo", "walrus"]), [
```
Honestly, I don't actually remember why I went with propEqual here. It [looks like](https://api.qunitjs.com/assert/propEqual) the differences between it and deepEqual mostly have to do with actual JS objects (i.e. not basic arrays like the ones here)—I can totally switch these asserts over to deepEqual if you'd like.

fedarko replied:
For consistency's sake, the tests now use `assert.deepEqual`.

fedarko committed:
STY: Change return signature to "explicit tuple"

Addresses @AmandaBirmingham's comment at
https://github.com/biocore/LabControl/pull/585#discussion_r322469591

fedarko commented on labcontrol/db/process.py:
```diff
@@ -298,6 +298,8 @@ def update_well(self, row, col, content):
         str
             The new contents of the well
+        bool
+            Whether or not the new contents of the well are "ok"
```
Understood and agreed! Just pushed a fix that changes this.

fedarko committed:
MAINT: Don't explicitly check limit in a handler

Instead, rely on Study.samples() to validate it.

Addresses @AmandaBirmingham's comment at
https://github.com/biocore/LabControl/pull/585#discussion_r322478733.

fedarko commented on labcontrol/gui/static/js/labcontrol.js:
```diff
@@ -742,3 +742,36 @@ function createHeatmap(
+  if (queryString.length === 0 || stringArray.length === []) {
+    return [];
+  }
+  var queryStringLowerCase = queryString.toLowerCase();
```
I'm ok with that (so long as we're ok with only supporting ES6-capable browsers). To clarify, do you just mean using `let` instead of `var` from here on, or actively porting the old code to use `let`?

`grep -ri "var " labcontrol/gui/static/js/ labcontrol/gui/js_tests/*.js | wc -l` finds 171 hits and `grep -ri "let " labcontrol/gui/static/js/ labcontrol/gui/js_tests/*.js | wc -l` finds 28 hits, so the latter option might take some time :)

fedarko committed:
DOC: updates -> update

fedarko committed:
DOC: Fix error in patchWell studyID description

This was due to my misunderstanding of how multiple studies are
handled in LC, as is documented in #437 :)

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -436,16 +434,18 @@ PlateViewer.prototype.getActiveStudy = function() {
  * @param {int} col The column of the well being modified
  * @param {string} content The new content of the well
+ * @param {string} studyID The output of getActiveStudy() -- this paradigm
+ *      should really be changed in the future to accommodate multiple studies.
```
Yeah—I wrote this comment before I realized that my understanding of how LabControl handles multiple studies was inaccurate. Just pushed a commit that should address this, thanks!

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -479,6 +483,57 @@ PlateViewer.prototype.modifyWell = function(row, col, content) {
+      } else {
+        safeArrayDelete(that.wellClasses[row][col], "well-indeterminate");
+      }
+      that.patchWell(row, col, possiblyNewContent, studyID);
```
`possiblyNewContent` might have been a bad choice for the variable name, sorry—even if this code fails to find a one-to-one match, it still updates the database with the `content` that was passed in (which also should be "new," since `modifyWell()` is called when a cell changes). I'm open to renaming this / adding comments to improve readability.

fedarko replied:
Renamed `possiblyNewContent` to `possiblyMatchedContent`, in the hopes of making this a bit clearer. Can do more if needed.

fedarko replied:
Renamed `possiblyNewContent` to `possiblyMatchedContent`, in the hopes of making this a bit clearer. Can do more if needed.

fedarko committed:
BUG: remove duplicate hidden IDs in plate template

fedarko commented on labcontrol/gui/templates/plate.html:
```diff
@@ -206,15 +206,19 @@ <h4>Plate map</h4>
             </div>
           </section>
           <!-- Hidden section added to populate a full row; otherwise, Well
               comments will not be left-justified. -->
```
I hadn't noticed the duplicate IDs, but I agree that that is egregious. Just pushed a commit that should fix that particular wart, at least.

It looks like duplicating the text is needed (at least with the way the HTML/CSS is written currently) because truncating the hidden text messes with the layout of the next section ("Well Comments")—

<img width="1287" alt="Screen Shot 2019-09-09 at 7 21 20 PM" src="https://user-images.githubusercontent.com/4177727/64579342-66086900-d337-11e9-94f7-6b6fadb78704.png">

For comparison, here's a screenshot with the duplicated text solution in use—

<img width="1219" alt="Screen Shot 2019-09-09 at 7 24 53 PM" src="https://user-images.githubusercontent.com/4177727/64579377-8801eb80-d337-11e9-81dd-78d4d0ab18d8.png">

I changed the hidden divs to be visible for these screenshots, but the effects shown here persist regardless of visibility.

Anyway! Yeah, entirely agreed that the need for this hidden section at all is a hack (and I hereby take responsibility for that, since I approved this code when it was merged in :). Removing the duplicate IDs makes it more tolerable, but it'd definitely be a good idea to adjust this to make the bottom row expand without needing a hidden section. I just created #590 to address this (can try to fix that in this PR if you'd like, but I don't think it's super urgent).

fedarko commented:
@AmandaBirmingham Thank you so much for the many helpful comments! I'm not going to be able to respond to them all today, but I'll get to work on hopefully addressing all of these tomorrow.

fedarko committed:
MAINT: Make study limit-checking more pythonic

That is, this just tries to see if int(limit) works successfully.

I also overhauled the tests accordingly, and accounted for a few
silly corner cases (e.g. someone passes in float('inf') as a limit).

This should address a few of @AmandaBirmingham's comments in #585.

fedarko commented on labcontrol/db/study.py:
```diff
@@ -222,7 +227,11 @@ def samples(self, term=None, limit=None):
+            if type(limit) != int:
+                raise ValueError("limit must be an int")
+            if limit <= 0:
+                raise ValueError("limit can't be <= 0")
```
Should be resolved in latest commit.

fedarko committed:
TST: test negative float limit case from handler

fedarko commented on labcontrol/db/tests/test_study.py:
```diff
@@ -83,6 +83,45 @@ def test_samples_with_sample_id(self):
+
+        # Check that non-int limits cause an error
+        nonint_limits_to_test = [0.01, 0.5, 1.0, 1.2, 3.0, 27.0, 29.1, 1000.0]
+        nonint_limits_to_test += [[1, 2, 3], "abc", "123", "1"]
```
Yep, now that `Study.samples()` accepts strings as input the `"1"` case would pass.

Something a little bit funky with the casting approach is that `int(1.0)` works (gets truncated to 1), but `int("1.0")` fails. Since (as far as I can tell) the only use of `Study.samples()` in LabControl involves passing in strings as inputs, this just means that if a request is sent with e.g. `...&limit=1.0` then that request will fail (as should be expected).

fedarko commented on labcontrol/gui/static/js/labcontrol.js:
```diff
@@ -742,3 +742,36 @@ function createHeatmap(
+ * This search is case insensitive, so "abc" will match with "ABCDEF".
+ *
+ * Note that if queryString is empty (i.e. "") then this won't return any
+ * matches, even if "" is present in stringArray for some reason (since
```
Thanks!

fedarko committed:
BUG/TST: catch TypeError on int(limit)

fedarko committed:
DOC: Fix description of getSubstringMatches return

Addresses a comment in #585

fedarko committed:
DOC: Remove comment that is superseded by #591

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -457,6 +457,10 @@ PlateViewer.prototype.modifyWell = function(row, col, content) {
+      // NOTE: this pretty much checks the status of every well, and this is
+      // called in patchWell (i.e. every time a well is updated). Should be a
+      // lot more efficient to minimize the impact of this by just checking for
+      // changes relative to this well?
```
Done: #591 now describes the issue.

fedarko committed:
DOC: Improve plateViewer docs/error messages

-Specify originating functions in the "new" error messages added
 to the plateViewer in #585
-Reword misleading sentence re: "assuming" in the docstring of
 merge_sample_responses()

Towards addressing @AmandaBirmingham's comments

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -745,26 +800,65 @@ function autocomplete_search_samples(request, response) {
+ * get_active_samples().
+ *
+ * NOTE that this assumes that the sample IDs in the response array are all
+ * unique. If for whatever reason a sample ID is repeated in a study -- or
```
Good catch, thank you!

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -789,6 +883,51 @@ function get_active_studies() {
+      },
+      function(jqXHR, textStatus, errorThrown) {
+        bootstrapAlert(
+          "Attempting to get a list of sample IDs failed: " +
```
Got it. I reworded the three uses of bootstrapAlerts in this commit to specify the function the alerts are being thrown from.

fedarko committed:
DOC: Copy a comment to where code has been copied

fedarko committed:
DOC: possiblyNewContent -> possiblyMatchedContent

fedarko committed:
DOC: add reference from comment to #592

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -479,6 +483,57 @@ PlateViewer.prototype.modifyWell = function(row, col, content) {
+  // batch operations.
+  var possiblyNewContent = content;
+  // TODO: cache list of active samples so that we don't have to make this
+  // particular request every time modifyWell() is called.
```
Created #592 accordingly.

fedarko committed:
DOC: Update modifyWell() docs re reviewer comments

fedarko commented on labcontrol/gui/static/js/plateViewer.js:
```diff
@@ -479,6 +483,57 @@ PlateViewer.prototype.modifyWell = function(row, col, content) {
+  // case could probably be detected in order to avoid doing matching here.)
+  //
+  // I expect the actual matching to be pretty fast (assuming that there
+  // aren't thousands of active samples); the main bottleneck is how requests
```
Understood—apologies for the naive comment. I've updated the comments accordingly, and now #173 is mentioned in this part of the code.

fedarko committed:
TST: For consistency, propEqual -> deepEqual

fedarko commented:
@AmandaBirmingham I think I've addressed all of your comments here (I left a few comments "unresolved" where I wasn't sure if you'd like me to do more).

Please let me know if you'd like me to do anything else before we merge this in—thanks again!

AmandaBirmingham commented on labcontrol/db/study.py:
```diff
@@ -228,10 +229,18 @@ def samples(self, term=None, limit=None):
+            # that the integer limit is greater than zero
+            try:
+                limit = int(limit)
+            except (ValueError, OverflowError, TypeError):
```
Is there a reason to limit this try/catch to particular errors?  Since the only statement wrapped in the try/catch is the cast, and since we aren't returning different error messages based on different error classes, I think it is safe to catch all errors (with a simple `except:`) and raise your custom ValueError.

fedarko replied:
Update—just pushed a commit that changes this to `except Exception:` (PEP8 apparently recommends this form over `except:`, since the latter catches stuff like KeyboardInterrupts).

AmandaBirmingham reviewed (approved):
Just one minor comment that I don't think requires action right now. Ready to merge after a second reviewer approves?

fedarko committed:
MAINT: More general error-handling of int(limit)

Addresses comment in #585

fedarko commented:
Since @gwarmstrong is out right now, I suppose I can request a review from someone else?

fedarko committed:
TST: for completeness' sake, test str(<= 0) limits

fedarko commented:
@qiyunzhu and I just sat down and walked through the code changes, and he has given his verbal approval of this PR.

While going over the code I noticed one small case that isn't explicitly tested in the `Study.samples()` unit test (strings of negative ints), so I added a check for that in the latest commit.

Anyway, once those tests pass (or at least fail only due to the intermittent and unrelated download error mentioned in #588), I can merge this PR in.

fedarko commented:
Build passed—merging in now. Thanks @AmandaBirmingham and @qiyunzhu for looking over this!
